src/simulators/state.hpp

```diff
@@ -344,9 +344,57 @@ void State<state_t>::apply_ops(InputIterator first, InputIterator last,
                                ExperimentResult &result,
                                RngEngine &rng,
                                bool final_ops) {
+
+  std::unordered_map<std::string, InputIterator> marks;
+  // Simple loop over vector of input operations
+  for (auto it = first; it != last; ++it) {
```

Comment:
This is also adding an unnecessary full iteration over all circuits just to find the jump locations. I think a more efficient way to handle this would be to add an unordered_map to `Circuit` that stores the mark locations, and this can be populated during the `set_params` method without an extra iteration, furthermore doing it this way would allow you to still apply truncation properly with these instructions since you can update their positions in the truncated circuit. IT will also allow you to catch errors for duplicate marks during circuit loading rather than execution.

Reply:
I will add check codes in `set_params`. This loop is not necessary because the next loop registers and looks-up `mark` instructions.